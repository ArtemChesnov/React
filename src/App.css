.App {
  text-align: center;
}

.App-logo {
  height: 40vmin;
  pointer-events: none;
}

@media (prefers-reduced-motion: no-preference) {
  .App-logo {
    animation: App-logo-spin infinite 20s linear;
  }
}

.App-header {
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
}

.scroll-btn {
  border: none;
  outline: none;
  background-color: transparent;
  padding: 2px 10px;
  color: #f8f8fa;
  font-size: 24px;
  text-transform: capitalize;
  cursor: pointer;
  transition: 0.15s;
}

.scroll-btn:hover {
  outline: 1px solid #f8f8fa;
  outline-offset: 10px;
  transition: all 200ms cubic-bezier(0.2, 0, 0, 0.8);
}

.scroll-btn:active {
  color: #282c34;
  outline-color: #282c34;
  outline-offset: 300px;
}

.App-main {
  background: #333842;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
}

.toDoList {
  height: 100vh;
  flex-direction: row-reverse;
  font-size: calc(10px + 2vmin);
  color: rgb(26, 25, 25);
  display: flex;
  justify-content: space-between;
  flex-direction: column;
  align-items: center;
}

.main-form {
  margin: 100px 0 0 0;
  display: flex;
  flex-direction: column;
}

.form-title {
  padding: 15px 25px;
  background-color: #454545;
  box-shadow: 2px 2px 3px rgb(255 255 255 / 10%);
  color: rgba(0, 0, 0, 0.6);
  text-shadow: 2px 2px 3px rgba(255, 255, 255, 0.1);
}

.form-wrp {
  display: flex;
}

.form-btn {
  font-size: 18px;
  cursor: pointer;
  transition: 0.3s;
  padding: 10px 15px;
  border: none;
  color: rgba(0, 0, 0, 0.6);
}

.form-btn:hover {
  background-color: rgb(141, 140, 140);
}

.form-box {
  position: relative;
}

.form-input {
  position: relative;
  margin: 0;
  border-radius: 0;
  font-size: 16px;
  outline: none;
  background-color: transparent;
  border: 1px solid rgb(87, 87, 87);
  width: 450px;
  height: 40px;
  padding: 0 0 0 10px;
  font-weight: bold;
  color: rgba(255, 255, 255, 0.6);
  text-shadow: 2px 2px 3px rgba(255, 255, 255, 0.1);
}

.form-input::placeholder {
  color: transparent;
}

.form-input:placeholder-shown~.form-label {
  font-size: 1.3rem;
  cursor: text;
  top: 20px;
}

.form-label {
  position: absolute;
  top: 0;
  left: 0;
  z-index: 0;
  transition: 0.2s;
  font-size: 14px;
  color: #9b9b9b;
  padding: 0 .2em;
  font-weight: bold;
  transform: translate(.5em, .8em);
  transform-origin: 0% 0%;
  background: #454545;
  pointer-events: none;
  transition: transform 0.3s ease, background 0.3s ease;
}

.form-input:focus+.form-label {
  transform: translate(.645rem, -.7em) scale(0.8);
  position: absolute;
  font-weight: 700;
}

.form-input:required,
.form-input:invalid {
  box-shadow: none;
}

.toDo-title {
  font-size: calc(20px + 2vmin);
  color: rgba(0, 0, 0, 0.6);
  margin: 0 0 15px 0;
}

.toDo-list {
  overflow: scroll;
  overflow-x: hidden;
  margin: 0;
  padding: 0;
  display: flex;
  flex-direction: column;
  text-align: start;
  max-height: 50vh;
  width: 100%;
  box-shadow: 2px 2px 3px rgb(255 255 255 / 10%);
}

.toDo-item {
  list-style-type: none;
  cursor: pointer;
  position: relative;
  padding: 12px 8px 12px 40px;
  background: #eee;
  font-size: 18px;
  transition: 0.2s;
  user-select: none;
}

.toDo-item:nth-child(odd) {
  background: #f9f9f9;
}

.toDo-item:hover {
  background: #ddd;
}

.toDo-item.checked {
  background: #888;
  color: #fff;
  text-decoration: line-through;
}

.toDo-item.checked::before {
  content: '';
  position: absolute;
  border-color: #fff;
  border-style: solid;
  border-width: 0 2px 2px 0;
  top: 10px;
  left: 16px;
  transform: rotate(45deg);
  height: 15px;
  width: 7px;
}

.close {
  position: absolute;
  right: 15px;
  top: 10px;
  transition: 0.2s ease-in-out;
}

.close:hover {
  transform: scale(1.8);
  color: rgb(37, 37, 37);
}

.parent {
  display: flex;
  flex-direction: column;
}

.scroll-btn-toDO {
  border: none;
  outline: none;
  background-color: transparent;
  padding: 2px 10px;
  color: #f8f8fa;
  font-size: 24px;
  text-transform: capitalize;
  cursor: pointer;
  transition: 0.15s;
  margin: 0 0 150px 0;
}

.scroll-btn-toDO:hover {
  outline: 1px solid #f8f8fa;
  outline-offset: 10px;
  transition: all 200ms cubic-bezier(0.2, 0, 0, 0.8);
}

.scroll-btn-toDO:active {
  color: #282c34;
  outline-color: #282c34;
  outline-offset: 300px;
}
