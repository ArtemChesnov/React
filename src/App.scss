.App {
  text-align: center;
}

.App-logo {
  height: 40vmin;
  pointer-events: none;
}

@media (prefers-reduced-motion: no-preference) {
  .App-logo {
    animation: App-logo-spin infinite 20s linear;
  }
}

.App-header {
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
}

.scroll-btn {
  border: none;
  outline: none;
  background-color: transparent;
  padding: 2px 10px;
  color: #f8f8fa;
  font-size: 24px;
  text-transform: capitalize;
  cursor: pointer;
  transition: 0.15s;
}

.scroll-btn:hover {
  outline: 1px solid #f8f8fa;
  outline-offset: 10px;
  transition: all 200ms cubic-bezier(0.2, 0, 0, 0.8);
}

.scroll-btn:active {
  color: #282c34;
  outline-color: #282c34;
  outline-offset: 300px;
}

.App-main {
  background: #333842;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  font-size: calc(10px + 2vmin);
  color: rgb(37, 37, 37);
}

.messageList {
  background: linear-gradient(90deg, rgba(0, 32, 36, 1) 0%, rgba(0, 255, 235, 0.9444152661064426) 100%);
  min-height: 100vh;
  display: flex;
}