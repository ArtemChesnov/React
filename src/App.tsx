import React, { FC, useMemo, useState } from 'react';
import { BrowserRouter, Route, Routes } from 'react-router-dom';
import { Header } from './Components/CompFunc/Header';
import { Chats } from './pages/Chats';
import { ChatList } from './Components/CompFunc/ChatList';
import { Home } from './pages/Home';
import { Profile } from './pages/Profile';
import { nanoid } from 'nanoid';
import style from './App.module.scss';

export interface Chat {
  id: string;
  name: string;
}

const initialMessage: Messages = {
  default: [
    {
      id: '1',
      author: 'GeekBrains',
      value: 'Hello, I`m geekbrains',
      now: new Date().toLocaleTimeString().slice(0, -3),
    },
  ],
};

export interface Message {
  id: string;
  author: string;
  value: string;
  now: string;
}

export interface Messages {
  [key: string]: Message[];
}

export const App: FC = () => {
  const [messages, setMessages] = useState<Messages>(initialMessage);

  const chatList = useMemo(
    () =>
      Object.entries(messages).map((chat) => ({
        id: nanoid(),
        name: chat[0],
      })),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [Object.entries(messages).length]
  );

  const onAddChat = (chat: Chat) => {
    if (!messages[chat.name]) {
      setMessages({
        ...messages,
        [chat.name]: [],
      });
    }
  };

  const onDeleteChat = (chatName: string) => {
    const newMessages: Messages = { ...messages };
    delete newMessages[chatName];

    setMessages({ ...newMessages });
  };

  return (
    <BrowserRouter>
      <Routes>
        <Route path="/" element={<Header />}>
          <Route index element={<Home />} />
          <Route path="profile" element={<Profile />} />

          <Route path="chats">
            <Route
              index
              element={
                <ChatList
                  chatList={chatList}
                  onAddChat={onAddChat}
                  onDeleteChat={onDeleteChat}
                />
              }
            />
            <Route
              path=":chatId"
              element={
                <Chats
                  messages={messages}
                  setMessages={setMessages}
                  chatList={chatList}
                  onAddChat={onAddChat}
                  onDeleteChat={onDeleteChat}
                />
              }
            />
          </Route>
        </Route>
        <Route
          path="*"
          element={
            <div className={style.err}>
              <div className={style.wrp}>
                <div className={style.circle}>
                  <h2 className={style.tittle}>404</h2>
                  <p className={style.text}>Упс...Что-то пошло не так...</p>
                </div>
              </div>
            </div>
          }
        />
      </Routes>
    </BrowserRouter>
  );
};
