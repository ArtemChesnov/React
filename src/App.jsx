import React, { useRef, useState, useCallback, useEffect } from 'react';
import './App.css';
import { Form } from './Components/CompFunc/Form';
import { Message } from './Components/CompFunc/Message';
import { MessageForm } from './Components/CompFunc/CompChat/MessageForm';
import { MessageList } from './Components/CompFunc/CompChat/MessageList/MessageList';

export const App = () => {
  const toToDo = useRef();
  const toChat = useRef();

  const [message, setMessage] = useState([]);

  const addMessage = useCallback(
    (autorValue, messageValue) => {
      setMessage([
        ...message,
        {
          autor: autorValue,
          value: messageValue,
          now: new Date().toLocaleTimeString().slice(0, -3),
        },
      ]);
    },
    [message]
  );

  useEffect(() => {
    if (message.length > 0 && message[message.length - 1].autor !== 'Душнила') {
      const timeout = setTimeout(() => {
        setMessage([
          ...message,
          {
            autor: 'Душнила',
            value: `Привет ${message[message.length - 1].autor}, ты скучный`,
            now: new Date().toLocaleTimeString().slice(0, -3),
          },
        ]);
      }, 2000);
      return () => {
        clearTimeout(timeout);
      };
    }
  }, [message]);

  return (
    <>
      <div className="App">
        <header className="App-header">
          <Message name={'Артём Чеснов'} />
          <button
            className="scroll-btn"
            onClick={() =>
              toToDo.current.scrollIntoView({ behavior: 'smooth' })
            }
          >
            Далее
          </button>
        </header>
        <main className="App-main">
          <section ref={toToDo} name={'toToDO'} className="toDoList">
            <Form />
            <button
              className="scroll-btn-toDO"
              onClick={() =>
                toChat.current.scrollIntoView({ behavior: 'smooth' })
              }
            >
              Далее
            </button>
          </section>
          <section ref={toChat} className="messageList">
            <MessageList message={message} />
            <MessageForm addMessage={addMessage} />
          </section>
        </main>
      </div>
    </>
  );
};
