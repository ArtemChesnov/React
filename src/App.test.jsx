import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import '@testing-library/user-event';
import { App } from './App';

describe('App', () => {
  it('Render component', () => {
    render(<App />);
  });

  it('Render with snapshot', () => {
    const { asFragment } = render(<App />);
    expect(asFragment()).toMatchSnapshot();
  });
});
