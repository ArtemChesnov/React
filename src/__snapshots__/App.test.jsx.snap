// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Render with snapshot 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <header
      class="App-header"
    >
      <h1>
        Привет!
      </h1>
      <h2>
        Меня зовут Артём Чеснов.
      </h2>
      <h3>
        И это мое первое приложение на React!
      </h3>
      <br />
      <br />
      <button
        class="scroll-btn"
      >
        Далее
      </button>
    </header>
    <main
      class="App-main"
    >
      <section
        class="toDoList"
        name="toToDO"
      >
        <form>
          <div
            class="main-form"
          >
            <div
              class="form-title"
            >
              <h1
                class="toDo-title"
              >
                Список дел
              </h1>
              <div
                class="form-wrp"
              >
                <div
                  class="form-box"
                >
                  <input
                    class="form-input"
                    required=""
                    type="input"
                    value=""
                  />
                  <label
                    class="form-label"
                  >
                    Название...
                  </label>
                </div>
                <button
                  class="form-btn"
                  type="button"
                >
                  Добавить
                </button>
              </div>
            </div>
            <div
              class="toDo-basket"
            >
              <ul
                class="toDo-list"
              />
            </div>
          </div>
        </form>
        <button
          class="scroll-btn-toDO"
        >
          Далее
        </button>
      </section>
      <section
        class="messageList"
      >
        <div
          class="message-board"
        >
          <h1
            class="message-title"
          >
            Чат с Душнилой
          </h1>
          <div
            class="message-chat"
          >
            <ul
              class="message-list"
            />
          </div>
        </div>
        <form
          data-testid="form"
        >
          <div
            class="mesageFromUser-form"
          >
            <div
              class="message-form"
            >
              <input
                class="autor-input"
                placeholder="Введите ваше имя..."
                required=""
                type="text"
                value=""
              />
              <textarea
                class="message-input"
                placeholder="Напишите сообщение..."
                required=""
                type="text"
              />
              <button
                class="message-button"
                disabled=""
                id="button"
                type="submit"
              >
                Отправить
              </button>
            </div>
          </div>
        </form>
      </section>
    </main>
  </div>
</DocumentFragment>
`;
